data-prepper-core/src/test/java/org/opensearch/dataprepper/peerforwarder/PeerForwarder_ClientServerIT.java

```diff
@@ -0,0 +1,280 @@
+                certificateProviderFactory, peerForwarderHttpService);
+
+        final Server server = serverProvider.get();
+
+        return new RemotePeerForwarderServer(peerForwarderConfiguration, server);
+    }
+
+    private PeerForwarderProvider createPeerForwarderProvider(
+            final PeerForwarderConfiguration peerForwarderConfiguration,
+            final CertificateProviderFactory certificateProviderFactory) {
+        final PeerForwarderClient clientForProvider = createClient(peerForwarderConfiguration, certificateProviderFactory);
+        final PeerClientPool peerClientPool = new PeerClientPool();
+        final PeerForwarderClientFactory clientFactoryForProvider = new PeerForwarderClientFactory(peerForwarderConfiguration, peerClientPool, certificateProviderFactory);
+        return new PeerForwarderProvider(clientFactoryForProvider, clientForProvider, peerForwarderConfiguration);
+    }
+
+    private PeerForwarderClient createClient(
+            final PeerForwarderConfiguration peerForwarderConfiguration,
+            final CertificateProviderFactory certificateProviderFactory) {
+        Objects.requireNonNull(peerForwarderConfiguration, "Nested classes must supply peerForwarderConfiguration");
+        Objects.requireNonNull(certificateProviderFactory, "Nested classes must supply certificateProviderFactory");
+        final PeerClientPool peerClientPool = new PeerClientPool();
+        final PeerForwarderClientFactory peerForwarderClientFactory = new PeerForwarderClientFactory(peerForwarderConfiguration, peerClientPool, certificateProviderFactory);
+        peerForwarderClientFactory.setPeerClientPool();
+        return new PeerForwarderClient(peerForwarderConfiguration, peerForwarderClientFactory, objectMapper);
+    }
+
+    @Nested
+    class WithSSL {
+
+        private PeerForwarderConfiguration peerForwarderConfiguration;
+        private CertificateProviderFactory certificateProviderFactory;
+        private PeerForwarderServer server;
+        private PeerForwarderProvider peerForwarderProvider;
+
+        @BeforeEach
+        void setUp() {
+            peerForwarderConfiguration = createConfiguration(true);
+
+            certificateProviderFactory = new CertificateProviderFactory(peerForwarderConfiguration);
+            peerForwarderProvider = createPeerForwarderProvider(peerForwarderConfiguration, certificateProviderFactory);
+            peerForwarderProvider.register(pipelineName, pluginId, Collections.singleton(UUID.randomUUID().toString()));
+            server = createServer(peerForwarderConfiguration, certificateProviderFactory, peerForwarderProvider);
+            server.start();
+        }
+
+        @AfterEach
+        void tearDown() {
+            server.stop();
+        }
+
+        @Test
+        void send_Events_to_server() {
+            final PeerForwarderClient client = createClient(peerForwarderConfiguration, certificateProviderFactory);
```

Comment:
Why are we creating client twice, we create it here and we also do in `setUp()` when we call `createPeerForwarderProvider`? Can we use same client?

Reply:
We could _possibly_ use the same client. But, I don't want to couple them. For example, I hope to create tests that use incorrect mTLS settings in the client. I don't want that affecting the server.